### PR TITLE
BlockToolbar: Show Group button in toolbar when multiple blocks are selected

### LINKED ISSUE #34461: Show "group" and variations (Row, Stack) in multi-select toolbar
Let's surface the "group" action on the multi-select toolbar:

image

It's a more common action to do when you multi-select and it's unnecessary buried:

image

## To-do

* [x] https://github.com/WordPress/gutenberg/pull/39710
* [ ] Follow up with additional buttons for Row and Stack variations

### PR DESCRIPTION
**Pull request by andrewserong**
## What?

Part of #34461

In this PR we expose a "Group" control in the block toolbar when multiple blocks are selected.

This PR borrows the implementation from #37619 (kudos @jiteshdhamaniya for opening that PR and kicking off this work! This PR is just a few tweaks on top of that one.)

## Why?

As raised in #34461, it will be helpful for users working with grouping together multiple blocks to have the control more visible and available in the block toolbar. After this PR lands, we could extend the approach to also include grouping multiple blocks into other containers e.g. Row and Cover blocks.

## How?

Add a `BlockGroupToolbar` component and conditionally render it in the block toolbar if multiple blocks are selected. If blocks cannot be grouped, we don't render that toolbar group.

## Testing Instructions

1. Insert a few blocks into a post. Shift + select the blocks and you should see the Group icon in the block toolbar.
2. Click it, and the blocks should now be grouped, and the group block should be selected.
3. In block editor preferences, switch to the "Display button label" toggle on, and check that the Group button in the block toolbar is displayed via the "Group" label text.
4. To ensure that the control _doesn't_ render when it shouldn't create a Social Icons block and add a couple of icons to the block. Select these child blocks together, and you should not see the Group option in the toolbar — this is because Group is not a valid child block of the Social Icons block.

## Screenshots or screencast 

| Group button in the block toolbar | Group button in display labels mode |
| --- | --- |
| image | image |

For reference, here is a screenshot of the Group button not being rendered when multiple social link blocks are selected:

### CODE REVIEW

**Comment by andrewserong** [comment]
I've just added a few folks as reviewers who've recently touched similar files — while this change is fairly small code wise, it's pretty visible, so I wanted to confidence check that the addition seems reasonable! 🙂

**Comment by Mamaduka** [comment]
A small note: we should probably hide the button if one of the blocks has removal locked. See #39541

**Comment by andrewserong** [comment]
Ah, perfect, thanks @Mamaduka, I'm glad I pinged you! 😀 I'll update.

**Comment by andrewserong** [comment]
I've made a similar change as in #39541 in https://github.com/WordPress/gutenberg/pull/39710/commits/d29a703bbe342ec91d474af50df25fdb3871b985, so blocks locked against removal should now prevent the Group button for showing. Cheers!

**Comment by Mamaduka** [comment]
Thank you, @andrewserong. I'll try to do full testing/review shortly.

**Comment by andrewserong** [comment]
No rush! I'm about to wrap up for the day. Thanks again for the 👀! 🙇

**Comment by jasmussen** [comment]
Nice, works well, I select multiple blocks and see this:
Screenshot 2022-03-24 at 10 24 45

This will be useful. 

Since we have a Row block that's a variant of the group already, and a Stack [underway](#39690), this could be a neat opportunity to both add unique icons for each. We could also surface these variants in the multi select toolbar:

Toolbar

For these group variants, we can also do better than the "Transform to variation" dropdown in the inspector, how about this?

Inspector

To be clear, I'm not suggesting all of this happen in this one single PR. But if we were to show both Row and Stack blocks next to the Group button, we'd need new icons for them anyway. What do you think?

_Speaking of the icons, I'm still tinkering a bit with the ones shown above. Work in progress!_

**Review by Mamaduka — APPROVED** [review]
It works as expected. Group button only visible if selected blocks can be grouped.

Thanks, @andrewserong.

File: packages/block-editor/src/components/convert-to-group-buttons/toolbar.js
Comment (Mamaduka) [inline packages/block-editor/src/components/convert-to-group-buttons/toolbar.js]: We might want to provide context for the label - `_x( 'Group', 'verb' )`. The `ConvertToGroupButton` also does this.

File: packages/block-editor/src/components/convert-to-group-buttons/toolbar.js
Comment (mtias) [inline packages/block-editor/src/components/convert-to-group-buttons/toolbar.js:55]: Let's add the variations of Group here as well — Row, Stack.

**Comment by mtias** [comment]
Looking good! Let's definitely try the variations of Group block as well with those icons. We can do it in follow ups if necessary.

File: packages/block-editor/src/components/convert-to-group-buttons/toolbar.js
Comment (andrewserong) [reply packages/block-editor/src/components/convert-to-group-buttons/toolbar.js:55]: Yes — I thought I'd look into that in a follow up PR. At the moment we have the same icon, so there's a couple ways we could do it, either adding additional icons in the row, or turn the Group button into a dropdown where you can then select which of the variations you'd like to convert to.

I'll have a play after this PR lands, and ping everyone again 😄

Edit: once Joen's neat new icons are ready, it looks like they'll be perfect for the toolbar 👍

**Comment by andrewserong** [comment]
Thanks for the feedback and designs @jasmussen, they look terrific!

I'll try to land this PR with just the Group block icon in the shorter-term, and then do a follow up adding in the other variations once the icons are ready. Do you want to ping me once you've got the icons ready, and I can open up a follow-up?

> For these group variants, we can also do better than the "Transform to variation" dropdown in the inspector, how about this?

I was just thinking yesterday that the drop-down isn't very intuitive, so I _love_ that button approach for switching variations — again, I'd be very happy to work on trying to implement that!

File: packages/block-editor/src/components/convert-to-group-buttons/toolbar.js
Comment (andrewserong) [reply packages/block-editor/src/components/convert-to-group-buttons/toolbar.js]: Good catch! Done in https://github.com/WordPress/gutenberg/pull/39710/commits/9e5c2132df956aee0f78ee1ad8bf2e3b5ee5bca2

**Comment by andrewserong** [comment]
I think this one's ready to land now. E2E tests are failing on `trunk` (and therefore this PR), but it looks like they should be resolved by https://github.com/WordPress/gutenberg/pull/39725. I'm wrapping up for the week in a little bit, but I'll aim to merge this one on Monday if no-one beats me to it!

**Comment by andrewserong** [comment]
Just managed to rebase this so that tests pass before I finish up for the week, so I'll merge this in now. Will follow up next week with looking into adding the remaining buttons. Thanks, all!

**Comment by mtias** [comment]
Great update, small but impactful :)

**Comment by jiteshdhamaniya** [comment]
Great work @andrewserong and thanks for the mention.  i will close my [PR](https://github.com/WordPress/gutenberg/pull/37619) now.

**Comment by andrewserong** [comment]
> For these group variants, we can also do better than the "Transform to variation" dropdown in the inspector, how about this?

I thought I'd have a go at this in a follow-up, so I've opened up a PR #40036 to explore adding in buttons for the Transform to Variation when each variation has a unique icon (the PR falls back to the dropdown menu if each variation does not have a unique icon).

**Pull request merged**
